src/workerd/api/gpu/gpu-buffer.c++

```diff
@@ -60,4 +85,83 @@ void GPUBuffer::unmap() {
   }
 }
 
+jsg::Promise<void> GPUBuffer::mapAsync(GPUFlagsConstant mode,
+                                       jsg::Optional<GPUSize64> offset,
+                                       jsg::Optional<GPUSize64> size) {
+  wgpu::MapMode md = static_cast<wgpu::MapMode>(mode);
+
+  if (state_ != State::Unmapped) {
+    device_.InjectError(
+        wgpu::ErrorType::Validation,
+        "mapAsync called on buffer that is not in the unmapped state");
+    JSG_FAIL_REQUIRE(
+        Error, "mapAsync called on buffer that is not in the unmapped state");
+  }
+
+  uint64_t o = 0;
+  KJ_IF_MAYBE (real_offset, offset) {
+    o = *real_offset;
+  }
+
+  uint64_t s = desc_.size - o;
+  KJ_IF_MAYBE (real_size, size) {
+    s = *real_size;
+  }
+
+  struct Context {
+    kj::Own<kj::PromiseFulfiller<void>> fulfiller;
+    State &state;
+    AsyncTask task;
+  };
+  auto paf = kj::newPromiseAndFulfiller<void>();
+  // This context object will old information for the callback, including the
+  // fullfiller to signal the caller with the result, and an async task that
+  // will ensure the device's Tick() function is called periodically. It will be
+  // deallocated at the end of the callback function.
+  auto ctx = new Context{kj::mv(paf.fulfiller), state_, AsyncTask(async_)};
+
+  state_ = State::MappingPending;
+
+  buffer_.MapAsync(
+      md, o, s,
+      [](WGPUBufferMapAsyncStatus status, void *userdata) {
+        auto c = std::unique_ptr<Context>(static_cast<Context *>(userdata));
+        c->state = State::Unmapped;
+
+        JSG_REQUIRE(c->fulfiller->isWaiting(), TypeError,
+                    "async operation has been canceled");
+
+        switch (status) {
+        case WGPUBufferMapAsyncStatus_Force32:
+          KJ_UNREACHABLE
+        case WGPUBufferMapAsyncStatus_Success:
+          c->fulfiller->fulfill();
+          c->state = State::Mapped;
+          break;
+        case WGPUBufferMapAsyncStatus_OffsetOutOfRange:
+        case WGPUBufferMapAsyncStatus_SizeOutOfRange:
+        case WGPUBufferMapAsyncStatus_ValidationError:
+          c->fulfiller->reject(KJ_EXCEPTION(FAILED, "validation failed"));
```

Comment:
These would surface as user-facing exceptions/rejections, correct? If so, these should be using `JSG_KJ_EXCEPTION` to construct a proper JS exception.